_posts/2019-01-XX-meltdown-and-spectre-vulnerabilities.md

```diff
@@ -0,0 +1,150 @@
+before the trap occurs the out-of-order execution takes place. As a result,
+instructions that follow the illegal access (causing the trap) are executed
+before being terminated. These instructions leak the contents of the accessed
+memory to the cache.
+
+## Variant 3a - rogue system register read
+
+This variant is also very similar to variant 3, but the target is different.
+In the variant 3 attacker wanted to obtain kernel memory space, but in this
+variant, hardware registers are accessed from user-space. Typically accessing
+hardware registers is restricted to root or even kernel itself. It is not
+possible to read some registers from user-space. However, exploiting the
+out-of-order execution allow leaking the register's information disclosing
+system parameters to an attacker with local user access.
+
+## Variant 4 - speculative store bypass
+
+Variant 4, when exploited, allows an attacker to read older memory values in a
+processor's stack or other memory locations. The trick is to make storing a
+value dependent on the results of previous instructions. This means that the
+processor has to wait before it knows where to store the value. The second
+step, the load, is, in contrast, constructed in such a way that the address
+can be determined quickly, without waiting. In this situation, the processor's
+speculative execution will "ignore" or "bypass" the store (because it doesn't
+yet know where the value is actually being stored) and just assume that the
+data currently held at the memory location is valid (but it's not).
+
+The processor will figure out its error, discard the results and perform
+correct calculations. But at this point, the microarchitectural state of the
+processor has already been changed. These changes can be detected, and an
+attacker can use those changes to figure out which value was read.
+
+# Summary
+
+For the past decades, we observed the increase in the processor's speed.
+Performance advancement has been achieved not only by the frequency itself but
+also by introducing speculative execution. Rather than wasting clock cycles to
+access memory or determine a value, CPU tries to predict the flow or value and
+executes instruction prematurely. If the guess was correct, the performance
+is increased, but if it was not correct, the CPU has to revert all changes to
+the last checkpoint (before incorrect execution) resulting in performance drop.
+
+In the next post, I will show You whether these vulnerabilities are exploitable
+on PC Engines apu2 platform and how to mitigate. So stay tuned.
+
+I hope this post was useful for you. Please feel free to share your opinion and
+if you think there is value, then share with friends.
+
+If you think we can help in improving the security of your firmware or you
+looking for someone who can boot your product by leveraging advanced features
+of used hardware platform, feel free to drop us email to `contact<at>3mdeb.com`
```

Comment:
@miczyg1 please improve that as we discussed in CPB `feel free to [boot a call with us](https://calendly.com/3mdeb/consulting-remote-meeting) or drop us email to ... If you are interested in similar content feel free to [sing up to our newsletter](http://eepurl.com/gfoekD)`

Reply:
Done